### DISCUSSION TITLE
TypeScript generated types make autoincrement required

### DISCUSSION THREAD

**Discussion by Krabba** [prisma/prisma#12703] [Q&A] [answered]
Hi,

For example, an ID field in my model that will autoincrement:
```
id                    Int                    @id @default(autoincrement())
```

Will give me this type in `node_modules\.prisma\client\index.d.ts`:
```
id: number
```

What I want to receive is:
```
id?: number
```

Should this not be optional since if you skip to provide it, it will auto increment if it's not defined in the Client API?

**Comment by austincrim** [comment, marked as answer]
Hey @Krabba 👋,

The model type itself will say that `id` is required unless it is explicitly marked as optional in your schema `id Int?`.

However, Prisma also generates separate types for creating or updating a model, and these types will have `id` marked as optional.

For this `User` model,
```prisma
model User {
  id   Int     @id @default(autoincrement())
  name String?
}
```
Prisma will generate types like these:
```ts
  export type UserUncheckedCreateInput = {
    id?: number
    name?: string | null
  }
```

Let me know if I can clarify any further!

**Comment by Krabba** [comment]
Hi @austincrim

You can't, or I can't atleast, mark an `@id` field as optional with `?`.
If you do that, you will get this error:
![image](https://user-images.githubusercontent.com/11424371/162383869-0201f6a1-889d-4360-aa55-d8c4cb00a48f.png)

However, I wasn't aware of the GIGANTIC variety of types to choose from when using the generated types in TypeScript.
I'll look in to those and let you know if I have anything else to ask.

Cool stuff!

Thank you for your help. 🛩️

**Comment by manojmpc91** [comment]
Well the problem still exists. 

The autoincrement fields are made mandatory by typescript and yes one cannot make those optional with ? either.

Does this have any solution yet?

**Comment by Krabba** [comment]
I am no longer able to, or was never able to import `UncheckedCreateInput` types from the `node_modules` folder.
Although i've located them and see that they do exist there.

I can only import my table name types from that file (which is weird), but not the others surrounding. 

I'm using `Prisma^4.1.1` as of now but always update to the latest version asap.

**Reply by austincrim** [reply]
Most of the generated types are under the `Prisma` namespace export.

```ts
import { Prisma } from '@prisma/client'

const input: Prisma.UserUncheckedCreateInput
```

**Comment by pablojsx** [comment]
Same issue but with uuid

**Comment by 10xchs** [comment]
Anyone with any solution to this? Preferably not austincrim who can't string a coherent sentence together if his life depended on it.

**Comment by mramazzini** [comment]
You can use Prisma.userCreateInput for a user model

This wont require the id field.

You can do this for any Prisma created model. 

So say you had a model FizzBuzz, the type Prisma.FizzBuzzCreateInput will be genereated.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
